ratelimits/limiter.go

```diff
@@ -0,0 +1,232 @@
+
+	// source is used to store buckets. It must be safe for concurrent use.
+	source source
+	clk    clock.Clock
+}
+
+// NewLimiter returns a new *Limiter. The provided source must be safe for
+// concurrent use. The defaults and overrides paths are expected to be paths to
+// YAML files that contain the default and override limits, respectively. The
+// overrides file is optional, all other arguments are required.
+func NewLimiter(clk clock.Clock, source source, defaults, overrides string) (*Limiter, error) {
+	limiter := &Limiter{source: source, clk: clk}
+
+	var err error
+	limiter.defaults, err = loadAndParseDefaultLimits(defaults)
+	if err != nil {
+		return nil, err
+	}
+
+	if overrides == "" {
+		// No overrides specified, initialize an empty map.
+		limiter.overrides = make(limits)
+		return limiter, nil
+	}
+
+	limiter.overrides, err = loadAndParseOverrideLimits(overrides)
+	if err != nil {
+		return nil, err
+	}
+
+	return limiter, nil
+}
+
+type Decision struct {
+	// Allowed is true if the bucket possessed enough capacity to allow the
+	// request given the cost.
+	Allowed bool
+
+	// Remaining is the number of requests the client is allowed to make before
+	// they're rate limited.
+	Remaining int
+
+	// RetryIn is the duration the client MUST wait before they're allowed to
+	// make a request.
+	RetryIn time.Duration
+
+	// ResetIn is the duration the bucket will take to refill to its maximum
+	// capacity, assuming no further requests are made.
+	ResetIn time.Duration
+
+	// newTAT is the next theoretical arrival time (TAT) of the next possible
+	// request.
+	newTAT time.Time
+}
+
+// Check returns a *Decision that indicates whether there's enough capacity to
+// allow the request given the cost, for the specified limit Name and client id.
+// However, it DOES NOT deduct the cost of the request from the bucket's
+// capacity. Hence, the returned *Decision represents the hypothetical state of
+// the bucket if the cost WERE to be deducted. The returned *Decision will
+// always include the number of remaining requests in the bucket, the required
+// wait time before the client can make another request, and the time until the
+// bucket refills to its maximum capacity (resets). If no bucket exists for the
+// given limit Name and client id, a new one will be created WITHOUT the
+// request's cost deducted from its initial capacity.
+func (l *Limiter) Check(name Name, id string, cost int64) (*Decision, error) {
+	if cost < 0 {
+		return nil, ErrInvalidCostForCheck
+	}
+
+	limit, err := l.getLimit(name, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if cost > limit.Burst {
+		return nil, ErrInvalidCostOverLimit
+	}
+
+	tat, err := l.source.Get(bucketKey(name, id))
+	if err != nil {
+		if !errors.Is(err, ErrBucketNotFound) {
+			return nil, err
+		}
+		// First request from this client.
+		d, err := l.initialize(limit, name, id, 0)
+		if err != nil {
+			return nil, err
+		}
+		return maybeSpend(l.clk, limit, d.newTAT, cost), nil
```

Comment:
Why initialize with a cost of zero, then spend with the actual cost, rather than just initializing with the actual cost?

Reply:
```
// Check returns a *Decision that indicates whether there's enough capacity to
// allow the request given the cost, for the specified limit Name and client id.
// However, it DOES NOT deduct the cost of the request from the bucket's
// capacity. Hence, the returned *Decision represents the hypothetical state of
// the bucket if the cost WERE to be deducted. The returned *Decision will
// always include the number of remaining requests in the bucket, the required
// wait time before the client can make another request, and the time until the
// bucket refills to its maximum capacity (resets). If no bucket exists for the
// given limit Name and client id, a new one will be created WITHOUT the
// request's cost deducted from its initial capacity.
```

As stated in the comment for this method:

> `Check returns a *Decision that indicates whether there's enough capacity to allow the request given the cost, for the specified limit Name and client id. However, it DOES NOT deduct the cost of the request from the bucket's capacity.`

> `If no bucket exists for the given limit Name and client id, a new one will be created WITHOUT the request's cost deducted from its initial capacity.`

I've added the following comment above the `initialize` call to avoid further confusion:

> `First request from this client. The cost is not deducted from the initial capacity because this is only a check.`